### ISSUE TITLE
Incorrect normalization of auxiliary loss in OLMoE and GPT Oss

### ISSUE THREAD

**Issue by andresnowak** [huggingface/transformers#43688]
### System Info

- `transformers` version: 4.57.6
- Platform: macOS-15.7.3-arm64-arm-64bit
- Python version: 3.10.14
- Huggingface_hub version: 0.36.0
- Safetensors version: 0.7.0
- Accelerate version: not installed
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.9.1 (NA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Expected behavior

From what I understand I think the balancing loss used in OLMoE and GPT-Oss in the huggingface implementation has an error.

```py
 expert_attention_mask = (
            attention_mask[None, :, :, None, None]
            .expand((num_hidden_layers, batch_size, sequence_length, top_k, num_experts))
            .reshape(-1, top_k, num_experts)
            .to(compute_device)
        )

        # Compute the percentage of tokens routed to each experts
        tokens_per_expert = torch.sum(expert_mask.float() * expert_attention_mask, dim=0) / torch.sum(
            expert_attention_mask, dim=0
        )
```

Here in this code the formula being used is:

$$
\begin{aligned} 
f_i &= \frac{N}{T} \sum_{t=1}^{T} \mathbb{1} (s_{i,t} \in \text{Topk}(\{s_{j,t} \mid 1 \leqslant j \leqslant N\}, K)), \quad \sum^N_i f_i = K \\ 
P_i &= \frac{1}{T} \sum_{t=1}^{T} s_{i,t} 
\end{aligned}
$$

But this is incorrect as $f_i$ should be normalized by $K$, in Switch transformer this isn't done because the model uses a Top-k of 1, but if we see how deepseek does like in [Deepseek-MoE](https://arxiv.org/abs/2401.06066), we need to normalize $f_i$ by $K$, if we don't do this then $f_i$ and $P_i$ represent two different distributions, where $f_i$ will have $\frac{K}{N}$ and $P_i$ will have $\frac{1}{N}$. 
And for OLMoE they mention that they use megablocks implementation for the balancing and Grouped GEMM and in [megablocks](https://github.com/Muennighoff/megablocks/blob/4a25bc7b5665bcb9da93d72d5ad0c14d41e1a351/megablocks/layers/moe.py#L31) we can see that it uses the correct normalized formula by $K$.

$$
\begin{aligned} 
f_i &= \frac{N}{T K} \sum_{t=1}^{T} \mathbb{1} (s_{i,t} \in \text{Topk}(\{s_{j,t} \mid 1 \leqslant j \leqslant N\}, K)), \quad \sum^N_i f_i = 1 \\ 
P_i &= \frac{1}{T} \sum_{t=1}^{T} s_{i,t} 
\end{aligned}
$$

With this both $f_i$ and $P_i$ can represent the same distribution.

**Comment by jiosephlee** [comment]
This seems still relevant?

**Comment by martinjaggi** [comment]
aux losses are still wrong currently, because the PR has not been merged yet. 

i can confirm aux losses for olmoe, qwen3, deepseek are all too high by a factor of top_k
